src/WebJobs.Script.WebHost/Security/KeyManagement/SecretManager.cs

```diff
@@ -90,12 +90,26 @@ public async virtual Task<HostSecretsInfo> GetHostSecretsAsync()
                         _logger.LogDebug("Loading host secrets");
 
                         hostSecrets = await LoadSecretsAsync<HostSecrets>();
-                        if (hostSecrets == null)
+                        try
+                        {
+                            if (hostSecrets == null)
+                            {
+                                // host secrets do not yet exist so generate them
+                                _logger.LogDebug(Resources.TraceHostSecretGeneration);
+                                hostSecrets = GenerateHostSecrets();
+                                await PersistSecretsAsync(hostSecrets);
+                            }
+                        }
+                        catch (Exception ex)
                         {
-                            // host secrets do not yet exist so generate them
-                            _logger.LogDebug(Resources.TraceHostSecretGeneration);
-                            hostSecrets = GenerateHostSecrets();
-                            await PersistSecretsAsync(hostSecrets);
+                            _logger.LogDebug("Trying again to ensure potential race condition is handled correctly");
+                            hostSecrets = await LoadSecretsAsync<HostSecrets>();
+
+                            if (hostSecrets == null)
+                            {
+                                _logger.LogError(ex, "Failed to load host secrets on second attempt");
```

Comment:
Normally I would put this in a loop and perform the entire load -> persist -> catch/ try again over like 10 attempts. But if you believe this won't be necessary here that is fine. I do see how you can assume if an exception is thrown _something else_ persisted the secrets.

Can you think of any other scenarios here that would throw but the secrets remain un-persisted?

Reply:
There could be brief connection issues as well (as an example) but I do think that most of the SDKs we use for persistence already retry for those kinds of things internally.

@aishwaryabh and I chatted about this when she started this PR and I explained this as a "let's give this one more try" kind of error handling b/c we've seen evidence of these races in the past, and rather than giving up, let's see if maybe things actually worked.

Since this code has been around for so long, I'd suggest we do this "one more try" approach, then we can monitor for how often we fail to load on the second attempt (and what kind of errors we see).

Reply:
Left it as is for the one more try approach